File: captainhcg/py/4-median-of-two-sorted-arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        def find_k(arr1, arr2, kth):
            if len(arr1) > len(arr2):
                arr2, arr1 = arr1, arr2
            if not arr1:
                return arr2[kth]
            if kth == 0:
                return min(arr1[0], arr2[0])
            m = min(len(arr1)-1, (kth-1)/2)
            if kth > m:
                if arr1[m] >= arr2[m]:
                    return find_k(arr1, arr2[m+1:], kth - m - 1)
                else:
                    return find_k(arr1[m+1:], arr2, kth - m - 1)
            else:  # kth <= m
                if arr1[m] >= arr2[m]:
                    return find_k(arr1[:m], arr2, kth)
                else:
                    return find_k(arr1, arr2[:m], kth)

        lensum = len(nums1) + len(nums2)
        if lensum % 2 == 1:
            m = find_k(nums1, nums2, lensum >> 1)
            return float(m)
        else:
            l = find_k(nums1, nums2, lensum >> 1)
            r = find_k(nums1, nums2, (lensum >> 1) - 1)
            return float(l + r) / 2

File: captainhcg/go/4-median-of-two-sorted-arrays.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    l := len(nums1) + len(nums2)
    if l % 2 == 0{
        return float64(findKth(nums1, nums2, l/2) + findKth(nums1, nums2, l/2-1)) / 2
    } else {
        return float64(findKth(nums1, nums2, l/2))
    }
}

func findKth(nums1 []int, nums2 []int, kth int) int{
	if len(nums1) >  len(nums2){
		nums1, nums2 = nums2, nums1
	}
	if len(nums1) == 0{
		return nums2[kth]
	}
	if kth == 0{
		return min(nums1[0], nums2[0])
	}
	mid := min(len(nums1) - 1, (kth-1) / 2)
	if kth >= mid{
		if nums1[mid] < nums2[mid]{
			return findKth(nums1[mid+1:], nums2, kth - mid - 1)
		} else {
			return findKth(nums1, nums2[mid+1:], kth - mid - 1)
		}
	} else { // kth < mid
		if nums1[mid] < nums2[mid]{
			return findKth(nums1, nums2[:mid], kth)
		} else {
			return findKth(nums1[:mid], nums2, kth)
		}
	}
}

func min(n1, n2 int) int {
	if n1 > n2{
		return n2
	}
	return n1
}
